docs/configs-spec2006-tests/run_spec.py

```diff
@@ -0,0 +1,250 @@
+                        help = "Path to the disk image containing SPEC benchmarks")
+    parser.add_argument("cpu", type = str, help = "Name of the detailed CPU")
+    parser.add_argument("benchmark", type = str,
+                        help = "Name of the SPEC benchmark")
+    parser.add_argument("size", type = str,
+                        help = "Available sizes: test, train, ref")
+    parser.add_argument("-k", "--kernel", type = str,
+                        default = "linux-4.19.83/vmlinux-4.19.83",
+                        help = "Path to vmlinux")
+    parser.add_argument("-d", "--disk", type = str,
+                        default = "disk-image/spec2006/spec2006-image/spec2006",
+                        help = "Path to the disk image containing SPEC benchmarks")
+    parser.add_argument("-l", "--no-copy-logs", default = False,
+                        action = "store_true", help = "Not copy SPEC run logs to the host system; Logs are copied by default")
+    parser.add_argument("-z", "--no-listeners", default = False,
+                        action = "store_true", help = "Turn off ports; The ports are on by default")
+    args = parser.parse_args()
+    
+    return args
+
+def getDetailedCPUModel(cpu_name):
+    '''
+    Return the CPU model corresponding to the cpu_name.
+    '''
+    available_models = {"kvm": X86KvmCPU,
+                        "o3": DerivO3CPU,
+                        "atomic": AtomicSimpleCPU}
+    try:
+        available_models["FlexCPU"] = FlexCPU
+    except NameError:
+        # FlexCPU is not defined
+        pass
+    # https://docs.python.org/3/library/stdtypes.html#dict.get 
+    # dict.get() returns None if the key does not exist
+    return available_models.get(cpu_name)
+
+def getBenchmarkName(benchmark_name):
+    if benchmark_name.endswith("(base)"):
+        benchmark_name = benchmark_name[:-6]
+    return benchmark_name
+
+def create_system(linux_kernel_path, disk_image_path, detailed_cpu_model):
+    # create the system we are going to simulate
+    system = MySystem(kernel = linux_kernel_path,
+                      disk = disk_image_path,
+                      num_cpus = 1, # run the benchmark in a single thread
+                      no_kvm = False,
+                      TimingCPUModel = detailed_cpu_model)
+    
+    # For workitems to work correctly
+    # This will cause the simulator to exit simulation when the first work
+    # item is reached and when the first work item is finished.
+    system.work_begin_exit_count = 1
+    system.work_end_exit_count = 1
+
+    # set up the root SimObject and start the simulation
+    root = Root(full_system = True, system = system)
+
+    if system.getHostParallel():
+        # Required for running kvm on multiple host cores.
+        # Uses gem5's parallel event queue feature
+        # Note: The simulator is quite picky about this number!
+        root.sim_quantum = int(1e9) # 1 ms
+
+    return root, system
+
+
+def boot_linux():
+    '''
```

Comment:
Can you add a comment on what this function is doing?

I guess it's running until an exit event because we know we exit after boot as soon as the script starts running?

Also, all of these functions have the same code. You could just call `run_to_exit` a bunch of times. Why did you split them into different functions? (This isn't something you necessarily need to change. I'm just trying to understand your thought process.)

Reply:
Initially, I thought they would be different as I was thinking about including switching CPU, dumping stats, etc in those functions. However, I later found that hiding the above tasks outside of the main function not ideal, because they are not specific to booting process and running the benchmark. As a result, I moved the tasks to the main function.

It was tempting to just have one function than having three functions doing the same thing. However, I think that, even though we have different functions doing the same thing, the function names are documenting our thought process pretty well, and it's clear that our implementation is consistent to the tutorial. It makes the main function a lot clearer too with just a small overhead.

Reply:
summary: I'm a fan of functional programming ;)